app/src/main/java/io/github/trojan_gfw/igniter/exempt/data/ExemptAppDataManager.java

```diff
@@ -1,52 +1,100 @@
 package io.github.trojan_gfw.igniter.exempt.data;
 
+import android.content.Context;
 import android.content.pm.ApplicationInfo;
 import android.content.pm.PackageManager;
 import android.os.Build;
+import android.support.annotation.NonNull;
+import android.text.TextUtils;
 
-import org.json.JSONArray;
-import org.json.JSONException;
-
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
 import java.util.ArrayList;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 
-import io.github.trojan_gfw.igniter.common.os.CommonSP;
-
+/**
+ * Implementation of {@link ExemptAppDataSource}. This class reads and writes exempted app list in a
+ * file named {@link #EXEMPT_APP_LIST_FILE_NAME} under {@link Context#getFilesDir()}. The exempted app
+ * package names will be written line by line in the file.
+ * <br/>
+ * Example:
+ * <br/>
+ * com.google.playstore
+ * <br/>
+ * io.github.trojan_gfw.igniter
+ * <br/>
+ * com.android.something
+ */
 public class ExemptAppDataManager implements ExemptAppDataSource {
+    private static final String EXEMPT_APP_LIST_FILE_NAME = "exempt_list";
     private final PackageManager mPackageManager;
+    private final String mExemptAppListFilePath;
 
-    public ExemptAppDataManager(PackageManager packageManager) {
+    public ExemptAppDataManager(Context context) {
         super();
-        mPackageManager = packageManager;
+        mPackageManager = context.getPackageManager();
+        String dir = context.getFilesDir().getAbsolutePath();
+        if (dir.endsWith(File.separator)) {
+            mExemptAppListFilePath = dir + EXEMPT_APP_LIST_FILE_NAME;
```

Comment:
Please extract file path to Globals

Reply:
Roger.

Reply:
Fixed. Now `ExemptAppDataManager`'s constructor requires a parameter named `exemptAppListFilePath`, which is is extracted from `Globals.getExemptedAppListPath()`.